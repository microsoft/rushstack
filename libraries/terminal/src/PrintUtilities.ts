// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Text } from '@rushstack/node-core-library';
import type { ITerminal } from './ITerminal';

/**
 * A sensible fallback column width for consoles.
 *
 * @public
 */
export const DEFAULT_CONSOLE_WIDTH: number = 80;

/**
 * A collection of utilities for printing messages to the console.
 *
 * @public
 */
export class PrintUtilities {
  /**
   * Returns the width of the console, measured in columns
   */
  public static getConsoleWidth(): number | undefined {
    return process.stdout?.columns;
  }

  /**
   * Applies word wrapping.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param indent - The number of spaces to indent the wrapped lines, defaults to 0
   */
  public static wrapWords(text: string, maxLineLength?: number, indent?: number): string;
  /**
   * Applies word wrapping.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param linePrefix - The string to prefix each line with, defaults to ''
   */
  public static wrapWords(text: string, maxLineLength?: number, linePrefix?: string): string;
  /**
   * Applies word wrapping.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param indentOrLinePrefix - The number of spaces to indent the wrapped lines or the string to prefix
   * each line with, defaults to no prefix
   */
  public static wrapWords(text: string, maxLineLength?: number, indentOrLinePrefix?: number | string): string;
  public static wrapWords(
    text: string,
    maxLineLength?: number,
    indentOrLinePrefix?: number | string
  ): string {
    const wrappedLines: string[] = PrintUtilities.wrapWordsToLines(
      text,
      maxLineLength,
      indentOrLinePrefix as string | undefined // TS is confused by the overloads
    );
    return wrappedLines.join('\n');
  }

  /**
   * Applies word wrapping and returns an array of lines.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param indent - The number of spaces to indent the wrapped lines, defaults to 0
   */
  public static wrapWordsToLines(text: string, maxLineLength?: number, indent?: number): string[];
  /**
   * Applies word wrapping and returns an array of lines.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param linePrefix - The string to prefix each line with, defaults to ''
   */
  public static wrapWordsToLines(text: string, maxLineLength?: number, linePrefix?: string): string[];
  /**
   * Applies word wrapping and returns an array of lines.
   *
   * @param text - The text to wrap
   * @param maxLineLength - The maximum length of a line, defaults to the console width
   * @param indentOrLinePrefix - The number of spaces to indent the wrapped lines or the string to prefix
   * each line with, defaults to no prefix
   */
  public static wrapWordsToLines(
    text: string,
    maxLineLength?: number,
    indentOrLinePrefix?: number | string
  ): string[];
  public static wrapWordsToLines(
    text: string,
    maxLineLength?: number,
    indentOrLinePrefix?: number | string
  ): string[] {
    let linePrefix: string;
    switch (typeof indentOrLinePrefix) {
      case 'number':
        linePrefix = ' '.repeat(indentOrLinePrefix);
        break;
      case 'string':
        linePrefix = indentOrLinePrefix;
        break;
      default:
        linePrefix = '';
        break;
    }

    const linePrefixLength: number = linePrefix.length;

    if (!maxLineLength) {
      maxLineLength = PrintUtilities.getConsoleWidth() || DEFAULT_CONSOLE_WIDTH;
    }

    // Apply word wrapping and the provided line prefix, while also respecting existing newlines
    // and prefix spaces that may exist in the text string already.
    const lines: string[] = Text.splitByNewLines(text);

    const wrappedLines: string[] = [];
    for (const line of lines) {
      if (line.length + linePrefixLength <= maxLineLength) {
        wrappedLines.push(linePrefix + line);
      } else {
        const lineAdditionalPrefix: string = line.match(/^\s*/)?.[0] || '';
        const whitespaceRegexp: RegExp = /\s+/g;
        let currentWhitespaceMatch: RegExpExecArray | null = null;
        let previousWhitespaceMatch: RegExpExecArray | undefined;
        let currentLineStartIndex: number = lineAdditionalPrefix.length;
        let previousBreakRanOver: boolean = false;
        while ((currentWhitespaceMatch = whitespaceRegexp.exec(line)) !== null) {
          if (currentWhitespaceMatch.index + linePrefixLength - currentLineStartIndex > maxLineLength) {
            let whitespaceToSplitAt: RegExpExecArray | undefined;
            if (
              !previousWhitespaceMatch ||
              // Handle the case where there are two words longer than the maxLineLength in a row
              previousBreakRanOver
            ) {
              whitespaceToSplitAt = currentWhitespaceMatch;
            } else {
              whitespaceToSplitAt = previousWhitespaceMatch;
            }

            wrappedLines.push(
              linePrefix +
                lineAdditionalPrefix +
                line.substring(currentLineStartIndex, whitespaceToSplitAt.index)
            );
            previousBreakRanOver = whitespaceToSplitAt.index - currentLineStartIndex > maxLineLength;
            currentLineStartIndex = whitespaceToSplitAt.index + whitespaceToSplitAt[0].length;
          } else {
            previousBreakRanOver = false;
          }

          previousWhitespaceMatch = currentWhitespaceMatch;
        }

        if (
          previousWhitespaceMatch &&
          line.length + linePrefixLength - currentLineStartIndex > maxLineLength
        ) {
          const whitespaceToSplitAt: RegExpExecArray = previousWhitespaceMatch;

          wrappedLines.push(
            linePrefix +
              lineAdditionalPrefix +
              line.substring(currentLineStartIndex, whitespaceToSplitAt.index)
          );
          currentLineStartIndex = whitespaceToSplitAt.index + whitespaceToSplitAt[0].length;
        }

        if (currentLineStartIndex < line.length) {
          wrappedLines.push(linePrefix + lineAdditionalPrefix + line.substring(currentLineStartIndex));
        }
      }
    }

    return wrappedLines;
  }

  /**
   * Displays a message in the console wrapped in a box UI.
   *
   * @param message - The message to display.
   * @param terminal - The terminal to write the message to.
   * @param boxWidth - The width of the box, defaults to half of the console width.
   */
  public static printMessageInBox(message: string, terminal: ITerminal, boxWidth?: number): void {
    if (!boxWidth) {
      const consoleWidth: number = PrintUtilities.getConsoleWidth() || DEFAULT_CONSOLE_WIDTH;
      boxWidth = Math.floor(consoleWidth / 2);
    }

    const maxLineLength: number = boxWidth - 10;
    const wrappedMessageLines: string[] = PrintUtilities.wrapWordsToLines(message, maxLineLength);
    let longestLineLength: number = 0;
    const trimmedLines: string[] = [];
    for (const line of wrappedMessageLines) {
      const trimmedLine: string = line.trim();
      trimmedLines.push(trimmedLine);
      longestLineLength = Math.max(longestLineLength, trimmedLine.length);
    }

    if (longestLineLength > boxWidth - 2) {
      // If the longest line is longer than the box, print bars above and below the message
      // ═════════════
      //  Message
      // ═════════════
      const headerAndFooter: string = ` ${'═'.repeat(boxWidth)}`;
      terminal.writeLine(headerAndFooter);
      for (const line of wrappedMessageLines) {
        terminal.writeLine(` ${line}`);
      }

      terminal.writeLine(headerAndFooter);
    } else {
      // ╔═══════════╗
      // ║  Message  ║
      // ╚═══════════╝
      terminal.writeLine(` ╔${'═'.repeat(boxWidth - 2)}╗`);
      for (const trimmedLine of trimmedLines) {
        const padding: number = boxWidth - trimmedLine.length - 2;
        const leftPadding: number = Math.floor(padding / 2);
        const rightPadding: number = padding - leftPadding;
        terminal.writeLine(` ║${' '.repeat(leftPadding)}${trimmedLine}${' '.repeat(rightPadding)}║`);
      }
      terminal.writeLine(` ╚${'═'.repeat(boxWidth - 2)}╝`);
    }
  }
}
