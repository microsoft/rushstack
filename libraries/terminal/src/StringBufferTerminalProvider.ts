// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StringBuilder, Text } from '@rushstack/node-core-library';
import { type ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';
import { AnsiEscape } from './AnsiEscape';

/**
 * @beta
 */
export interface IStringBufferOutputOptions {
  /**
   * If set to true, special characters like \\n, \\r, and the \\u001b character
   * in color control tokens will get normalized to [-n-], [-r-], and [-x-] respectively
   *
   * This option defaults to `true`
   */
  normalizeSpecialCharacters: boolean;
}

/**
 * Terminal provider that stores written data in buffers separated by severity.
 * This terminal provider is designed to be used when code that prints to a terminal
 * is being unit tested.
 *
 * @beta
 */
export class StringBufferTerminalProvider implements ITerminalProvider {
  private _standardBuffer: StringBuilder = new StringBuilder();
  private _verboseBuffer: StringBuilder = new StringBuilder();
  private _debugBuffer: StringBuilder = new StringBuilder();
  private _warningBuffer: StringBuilder = new StringBuilder();
  private _errorBuffer: StringBuilder = new StringBuilder();

  private _supportsColor: boolean;

  public constructor(supportsColor: boolean = false) {
    this._supportsColor = supportsColor;
  }

  /**
   * {@inheritDoc ITerminalProvider.write}
   */
  public write(data: string, severity: TerminalProviderSeverity): void {
    switch (severity) {
      case TerminalProviderSeverity.warning: {
        this._warningBuffer.append(data);
        break;
      }

      case TerminalProviderSeverity.error: {
        this._errorBuffer.append(data);
        break;
      }

      case TerminalProviderSeverity.verbose: {
        this._verboseBuffer.append(data);
        break;
      }

      case TerminalProviderSeverity.debug: {
        this._debugBuffer.append(data);
        break;
      }

      case TerminalProviderSeverity.log:
      default: {
        this._standardBuffer.append(data);
        break;
      }
    }
  }

  /**
   * {@inheritDoc ITerminalProvider.eolCharacter}
   */
  public get eolCharacter(): string {
    return '\n';
  }

  /**
   * {@inheritDoc ITerminalProvider.supportsColor}
   */
  public get supportsColor(): boolean {
    return this._supportsColor;
  }

  /**
   * Get everything that has been written at log-level severity.
   */
  public getOutput(options?: IStringBufferOutputOptions): string {
    return this._normalizeOutput(this._standardBuffer.toString(), options);
  }

  /**
   * @deprecated - use {@link StringBufferTerminalProvider.getVerboseOutput}
   */
  public getVerbose(options?: IStringBufferOutputOptions): string {
    return this.getVerboseOutput(options);
  }

  /**
   * Get everything that has been written at verbose-level severity.
   */
  public getVerboseOutput(options?: IStringBufferOutputOptions): string {
    return this._normalizeOutput(this._verboseBuffer.toString(), options);
  }

  /**
   * Get everything that has been written at debug-level severity.
   */
  public getDebugOutput(options?: IStringBufferOutputOptions): string {
    return this._normalizeOutput(this._debugBuffer.toString(), options);
  }

  /**
   * Get everything that has been written at error-level severity.
   */
  public getErrorOutput(options?: IStringBufferOutputOptions): string {
    return this._normalizeOutput(this._errorBuffer.toString(), options);
  }

  /**
   * Get everything that has been written at warning-level severity.
   */
  public getWarningOutput(options?: IStringBufferOutputOptions): string {
    return this._normalizeOutput(this._warningBuffer.toString(), options);
  }

  private _normalizeOutput(s: string, options: IStringBufferOutputOptions | undefined): string {
    options = {
      normalizeSpecialCharacters: true,

      ...(options || {})
    };

    s = Text.convertToLf(s);

    if (options.normalizeSpecialCharacters) {
      return AnsiEscape.formatForTests(s, { encodeNewlines: true });
    } else {
      return s;
    }
  }
}
