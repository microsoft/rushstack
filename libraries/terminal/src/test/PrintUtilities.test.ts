// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { Terminal } from '../Terminal';
import { PrintUtilities } from '../PrintUtilities';

describe(PrintUtilities.name, () => {
  let terminalProvider: StringBufferTerminalProvider;
  let terminal: Terminal;

  beforeEach(() => {
    terminalProvider = new StringBufferTerminalProvider(false);
    terminal = new Terminal(terminalProvider);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  function testWrapWordsToLines(prefix: string | number | undefined): void {
    describe(`with prefix="${prefix}"`, () => {
      it('respects spaces and newlines in a pre-formatted message', () => {
        const userMessage: string = [
          'An error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
          '',
          '    common/scripts/setup.sh',
          ''
        ].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(userMessage, 50, prefix);
        expect(result).toMatchSnapshot();
      });

      it('handles a line starting with a word longer than the max line length', () => {
        const userMessage: string = [
          'Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
          '',
          '    common/scripts/setup.sh',
          ''
        ].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(userMessage, 50, prefix);
        expect(result).toMatchSnapshot();
      });

      it('handles a line containing a word longer than the max line length', () => {
        const userMessage: string = [
          'An error occurrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
          '',
          '    common/scripts/setup.sh',
          ''
        ].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(userMessage, 50, prefix);
        expect(result).toMatchSnapshot();
      });

      it('handles a line starting with two words longer than the max line length', () => {
        const userMessage: string = [
          'Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn errrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrror occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
          '',
          '    common/scripts/setup.sh',
          ''
        ].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(userMessage, 50, prefix);
        expect(result).toMatchSnapshot();
      });

      it('handles a line with only a word longer than the max line length', () => {
        const userMessage: string = ['Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn'].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(userMessage, 50, prefix);
        expect(result).toMatchSnapshot();
      });

      it('applies pre-existing indents on both margins', () => {
        const message: string = [
          'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.',
          '',
          '    Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.',
          '',
          'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.'
        ].join('\n');

        const result: string[] = PrintUtilities.wrapWordsToLines(message, 50, prefix);
        expect(result).toMatchSnapshot();
      });
    });
  }

  describe(PrintUtilities.wrapWordsToLines.name, () => {
    testWrapWordsToLines(undefined);
    testWrapWordsToLines(4);
    testWrapWordsToLines('| ');
  });

  describe(PrintUtilities.printMessageInBox.name, () => {
    function validateOutput(expectedWidth: number): void {
      const outputLines: string[] = terminalProvider
        .getOutput({ normalizeSpecialCharacters: true })
        .split('[n]');
      expect(outputLines).toMatchSnapshot();

      expect(outputLines.every((x) => x.length <= expectedWidth));
    }

    const MESSAGE: string =
      'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.';

    it('prints a long message wrapped in narrow box', () => {
      PrintUtilities.printMessageInBox(MESSAGE, terminal, 20);
      validateOutput(20);
    });

    it('prints a long message wrapped in a wide box', () => {
      PrintUtilities.printMessageInBox(MESSAGE, terminal, 300);
      validateOutput(300);
    });

    it('prints a long message wrapped in a box using the console width', () => {
      jest.spyOn(PrintUtilities, 'getConsoleWidth').mockReturnValue(65);

      PrintUtilities.printMessageInBox(MESSAGE, terminal);
      validateOutput(32);
    });

    it('respects spaces and newlines in a pre-formatted message', () => {
      const userMessage: string = [
        'An error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
        '',
        '    common/scripts/setup.sh',
        ''
      ].join('\n');

      PrintUtilities.printMessageInBox(userMessage, terminal, 50);
      validateOutput(50);
    });

    it('Handles a case where there is a word longer than the boxwidth', () => {
      const userMessage: string = [
        'Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
        '',
        '    common/scripts/setup.sh',
        ''
      ].join('\n');

      PrintUtilities.printMessageInBox(userMessage, terminal, 50);
      validateOutput(50);
    });

    it('word-wraps a message with a trailing fragment', () => {
      const lines: string[] = PrintUtilities.wrapWordsToLines(
        'This Thursday, we will complete the Node.js version upgrade.  Any pipelines that still have not upgraded will be temporarily disabled.',
        36
      );
      expect(lines).toMatchInlineSnapshot(`
Array [
  "This Thursday, we will complete the",
  "Node.js version upgrade.  Any",
  "pipelines that still have not",
  "upgraded will be temporarily",
  "disabled.",
]
`);

      for (const line of lines) {
        expect(line.length).toBeLessThanOrEqual(36);
      }
    });
  });
});
