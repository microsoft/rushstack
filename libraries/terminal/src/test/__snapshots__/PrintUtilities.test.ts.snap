// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrintUtilities printMessageInBox Handles a case where there is a word longer than the boxwidth 1`] = `
Array [
  " ══════════════════════════════════════════════════",
  " Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  " error occurred while pushing commits to",
  " git remote. Please make sure you have",
  " installed and enabled git lfs. The",
  " easiest way to do that is run the",
  " provided setup script:",
  " ",
  "     common/scripts/setup.sh",
  " ",
  " ══════════════════════════════════════════════════",
  "",
]
`;

exports[`PrintUtilities printMessageInBox prints a long message wrapped in a box using the console width 1`] = `
Array [
  " ╔══════════════════════════════╗",
  " ║    Lorem ipsum dolor sit     ║",
  " ║      amet, consectetuer      ║",
  " ║       adipiscing elit.       ║",
  " ║      Maecenas porttitor      ║",
  " ║     congue massa. Fusce      ║",
  " ║      posuere, magna sed      ║",
  " ║     pulvinar ultricies,      ║",
  " ║    purus lectus malesuada    ║",
  " ║       libero, sit amet       ║",
  " ║      commodo magna eros      ║",
  " ║          quis urna.          ║",
  " ╚══════════════════════════════╝",
  "",
]
`;

exports[`PrintUtilities printMessageInBox prints a long message wrapped in a wide box 1`] = `
Array [
  " ╔══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗",
  " ║                                              Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.                                               ║",
  " ╚══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝",
  "",
]
`;

exports[`PrintUtilities printMessageInBox prints a long message wrapped in narrow box 1`] = `
Array [
  " ╔══════════════════╗",
  " ║      Lorem       ║",
  " ║      ipsum       ║",
  " ║    dolor sit     ║",
  " ║      amet,       ║",
  " ║   consectetuer   ║",
  " ║ adipiscing elit. ║",
  " ║     Maecenas     ║",
  " ║    porttitor     ║",
  " ║      congue      ║",
  " ║      massa.      ║",
  " ║      Fusce       ║",
  " ║     posuere,     ║",
  " ║    magna sed     ║",
  " ║     pulvinar     ║",
  " ║    ultricies,    ║",
  " ║      purus       ║",
  " ║      lectus      ║",
  " ║    malesuada     ║",
  " ║     libero,      ║",
  " ║     sit amet     ║",
  " ║     commodo      ║",
  " ║    magna eros    ║",
  " ║    quis urna.    ║",
  " ╚══════════════════╝",
  "",
]
`;

exports[`PrintUtilities printMessageInBox respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  " ╔════════════════════════════════════════════════╗",
  " ║    An error occurred while pushing commits     ║",
  " ║    to git remote. Please make sure you have    ║",
  " ║       installed and enabled git lfs. The       ║",
  " ║       easiest way to do that is run the        ║",
  " ║             provided setup script:             ║",
  " ║                                                ║",
  " ║            common/scripts/setup.sh             ║",
  " ║                                                ║",
  " ╚════════════════════════════════════════════════╝",
  "",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " applies pre-existing indents on both margins 1`] = `
Array [
  "| Lorem ipsum dolor sit amet, consectetuer",
  "| adipiscing elit. Maecenas porttitor congue",
  "| massa.",
  "| ",
  "|     Lorem ipsum dolor sit amet, consectetuer",
  "|     adipiscing elit. Maecenas porttitor congue",
  "|     massa.",
  "| ",
  "| Lorem ipsum dolor sit amet, consectetuer",
  "| adipiscing elit. Maecenas porttitor congue",
  "| massa.",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " handles a line containing a word longer than the max line length 1`] = `
Array [
  "| An error",
  "| occurrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrred",
  "| while pushing commits to git remote. Please make",
  "| sure you have installed and enabled git lfs. The",
  "| easiest way to do that is run the provided setup",
  "| script:",
  "| ",
  "|     common/scripts/setup.sh",
  "| ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " handles a line starting with a word longer than the max line length 1`] = `
Array [
  "| Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "| error occurred while pushing commits to git",
  "| remote. Please make sure you have installed and",
  "| enabled git lfs. The easiest way to do that is",
  "| run the provided setup script:",
  "| ",
  "|     common/scripts/setup.sh",
  "| ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " handles a line starting with two words longer than the max line length 1`] = `
Array [
  "| Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "| errrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrror",
  "| occurred while pushing commits to git remote.",
  "| Please make sure you have installed and enabled",
  "| git lfs. The easiest way to do that is run the",
  "| provided setup script:",
  "| ",
  "|     common/scripts/setup.sh",
  "| ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " handles a line with only a word longer than the max line length 1`] = `
Array [
  "| Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="| " respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  "| An error occurred while pushing commits to git",
  "| remote. Please make sure you have installed and",
  "| enabled git lfs. The easiest way to do that is",
  "| run the provided setup script:",
  "| ",
  "|     common/scripts/setup.sh",
  "| ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" applies pre-existing indents on both margins 1`] = `
Array [
  "    Lorem ipsum dolor sit amet, consectetuer",
  "    adipiscing elit. Maecenas porttitor congue",
  "    massa.",
  "    ",
  "        Lorem ipsum dolor sit amet, consectetuer",
  "        adipiscing elit. Maecenas porttitor congue",
  "        massa.",
  "    ",
  "    Lorem ipsum dolor sit amet, consectetuer",
  "    adipiscing elit. Maecenas porttitor congue",
  "    massa.",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" handles a line containing a word longer than the max line length 1`] = `
Array [
  "    An error",
  "    occurrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrred",
  "    while pushing commits to git remote. Please",
  "    make sure you have installed and enabled git",
  "    lfs. The easiest way to do that is run the",
  "    provided setup script:",
  "    ",
  "        common/scripts/setup.sh",
  "    ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" handles a line starting with a word longer than the max line length 1`] = `
Array [
  "    Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "    error occurred while pushing commits to git",
  "    remote. Please make sure you have installed",
  "    and enabled git lfs. The easiest way to do",
  "    that is run the provided setup script:",
  "    ",
  "        common/scripts/setup.sh",
  "    ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" handles a line starting with two words longer than the max line length 1`] = `
Array [
  "    Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "    errrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrror",
  "    occurred while pushing commits to git remote.",
  "    Please make sure you have installed and",
  "    enabled git lfs. The easiest way to do that is",
  "    run the provided setup script:",
  "    ",
  "        common/scripts/setup.sh",
  "    ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" handles a line with only a word longer than the max line length 1`] = `
Array [
  "    Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="4" respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  "    An error occurred while pushing commits to git",
  "    remote. Please make sure you have installed",
  "    and enabled git lfs. The easiest way to do",
  "    that is run the provided setup script:",
  "    ",
  "        common/scripts/setup.sh",
  "    ",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" applies pre-existing indents on both margins 1`] = `
Array [
  "Lorem ipsum dolor sit amet, consectetuer",
  "adipiscing elit. Maecenas porttitor congue massa.",
  "",
  "    Lorem ipsum dolor sit amet, consectetuer",
  "    adipiscing elit. Maecenas porttitor congue massa.",
  "",
  "Lorem ipsum dolor sit amet, consectetuer",
  "adipiscing elit. Maecenas porttitor congue massa.",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" handles a line containing a word longer than the max line length 1`] = `
Array [
  "An error",
  "occurrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrred",
  "while pushing commits to git remote. Please make",
  "sure you have installed and enabled git lfs. The",
  "easiest way to do that is run the provided setup",
  "script:",
  "",
  "    common/scripts/setup.sh",
  "",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" handles a line starting with a word longer than the max line length 1`] = `
Array [
  "Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "error occurred while pushing commits to git",
  "remote. Please make sure you have installed and",
  "enabled git lfs. The easiest way to do that is run",
  "the provided setup script:",
  "",
  "    common/scripts/setup.sh",
  "",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" handles a line starting with two words longer than the max line length 1`] = `
Array [
  "Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
  "errrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrror",
  "occurred while pushing commits to git remote.",
  "Please make sure you have installed and enabled",
  "git lfs. The easiest way to do that is run the",
  "provided setup script:",
  "",
  "    common/scripts/setup.sh",
  "",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" handles a line with only a word longer than the max line length 1`] = `
Array [
  "Annnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn",
]
`;

exports[`PrintUtilities wrapWordsToLines with prefix="undefined" respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  "An error occurred while pushing commits to git",
  "remote. Please make sure you have installed and",
  "enabled git lfs. The easiest way to do that is run",
  "the provided setup script:",
  "",
  "    common/scripts/setup.sh",
  "",
]
`;
